upb/refcounted.c

```diff
@@ -726,6 +726,18 @@ static void freeobj(upb_refcounted *o) {
 bool upb_refcounted_init(upb_refcounted *r,
                          const struct upb_refcounted_vtbl *vtbl,
                          const void *owner) {
+#ifndef NDEBUG
+  // Endianness check.  This is unrelated to upb_refcounted, it's just a
+  // convenient place to put the check that we can be assured will run for
+  // basically every program using upb.
+  const int x = 1;
+#ifdef UPB_BIG_ENDIAN
+  assert(*(char*)&x != 1);
+#else
+  assert(*(char*)&x == 1);
```

Comment:
This will run on every descriptor object creation, right? Maybe not too big a deal, but I wonder if it could be made a compile-time check.

Actually, scratch that. gcc is apparently smart enough to do constant-prop through a typecast -- this compiles down to a `ret` instruction with `-O3`:
    void f() {
      const int x = 1;
      assert(*(char*)&x == 1);
    }

Maybe add a comment to allay equivalent worries from future readers that this should be optimized away?